### ISSUE TITLE
mpifileutils: error while compiling from  source 

### ISSUE THREAD

**Issue by Zubrania** [hpc/mpifileutils#108]
Hello  

Can  you please  look at the error 

In file included from mfu_flist.c:63:
/root/HPE-Lustre/mpifileutils/install/include/dtcmp.h:69: warning: function declaration isn't a prototype
/root/HPE-Lustre/mpifileutils/install/include/dtcmp.h:72: warning: function declaration isn't a prototype
In file included from mfu.h:48,
                 from mfu_flist.c:64:
mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
mfu_flist.c: In function 'list_elem_pack2_size':
mfu_flist.c:114: warning: unused parameter 'elem'
mfu_flist.c: In function 'mfu_flist_remap':
mfu_flist.c:1083: warning: passing argument 4 of 'map' discards qualifiers from pointer target type
mfu_flist.c:1083: note: expected 'void *' but argument is of type 'const void *'
mfu_flist.c: In function 'map_spread':
mfu_flist.c:1165: warning: unused parameter 'args'
make[3]: *** [libmfu_la-mfu_flist.lo] Error 1
make[3]: Leaving directory `/root/HPE-Lustre/mpifileutils/mpifileutils-0.7/src/common'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/root/HPE-Lustre/mpifileutils/mpifileutils-0.7/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/HPE-Lustre/mpifileutils/mpifileutils-0.7'
make: *** [all] Error 2
+ '[' 2 -ne 0 ']'
+ echo 'failed to configure, build, or install mpiFileUtils'
failed to configure, build, or install mpiFileUtils

I believe I am  missing some  dependences  but unfortunately  can not find out  which  ones?

Thank you

**Comment by dsikich** [comment]
@Zubrania Hi, it looks like you are missing the xattrs headers. Are you using Lustre? 

Can you send me the steps you used to build? Did you run the "buildme" script? If so, there is a -DDCOPY_USE_XATTRS in there that you can delete, and try to re-run. Also, did you run the "buildme_dependencies" script prior to trying to build with "buildme"?

**Comment by Zubrania** [comment]
Hello please  see my  answers  below


> On Mar 15, 2018, at 23:40, dsikich <notifications@github.com> wrote:
> 
> @Zubrania <https://github.com/zubrania> Hi, it looks like you are missing the xattrs headers. Are you using Lustre?
> 
> Can you send me the steps you used to build?
> 
I have tried to compile the code according to the documentation 
./buildme_dependencies

./buildme

without usage of experimental tools
> Did you run the "buildme" script?
> 
yes
> If so, there is a -DDCOPY_USE_XATTRS in there that you can delete, and try to re-run.
> 
I gave commented out the export CFLAGS="-g -O0 -I${topdir}/src/common -DDCOPY_USE_XATTRS”
and run buildme  again/ Finally  the build completed  without any  errors
> Also, did you run the "buildme_dependencies" script prior to trying to build with "buildme"?
> 
> 
sure  that  has been  done

**Comment by dsikich** [comment]
@Zubrania good to hear you were able to build without any errors. Were you able to run the tools successfully as well?

**Comment by Zubrania** [comment]
I am  looking  for a proper way to generate the list of files to be copied by  dcp. And unfortunately  did not find  yet

**Comment by Zubrania** [comment]
With  the openmpi version  change  I had to recompile  the  mpifileutils 

the buildme_dependecies run  smoothly  after uncommenting the 
# Open MPI
curdir=`pwd`
export PATH="${curdir}/openmpi/install/bin:/usr/lib64/openmpi/bin:$PATH"

and adding the mpicc path 

But while running buildme  I got the  error

checking for mpicc... no
checking for mpiicc... no
checking for mpixlc... no
checking for mpipgcc... no
Checking whether not-found responds to '-showme:compile'... no
Checking whether not-found responds to '-showme'... no
Checking whether not-found responds to '-compile-info'... no
Checking whether not-found responds to '-show'... no
Unable to find suitable MPI Compiler. Try setting MPICC.
configure: error: You should check to see if MPI is setup properly.
+ '[' 1 -ne 0 ']'
+ echo 'failed to configure, build, or install mpiFileUtils'
failed to configure, build, or install mpiFileUtils
+ exit 1
[root@node25 mpifileutils]# vim buildme

saying that mpicc is missing. Can you  please  advice  what needs to be done to  fix  that ?
Thank you

**Comment by Zubrania** [comment]
@dsikich

Yes I am planning using this tool to do datamigration  to lustre

the steps were the following
1) Editing buildme_dependencies 
	Uncommenting PATH and adding location  of mpi binaries

2) Run buildme_dependencies  which  produce no errors
3) Run  buildme   and getting  the error about missing mpicc, that I have provided earlier

environment - RHEL 6.9

**Comment by dsikich** [comment]
@Zubrania Also, it doesn't sound like you will be doing any new feature development. In that case, do you have spack (https://github.com/spack/spack/blob/develop/lib/spack/docs/index.rst). I think that will simplify your build/installation if you just want to run and use the tools. 

Once you've downloaded spack you should be able to just do a:

"spack install mpifileutils +lustre"

**Comment by Zubrania** [comment]
@dsikich
There still error appears while building using spack. from  what I see  it is somehow connected with -DDCOPY_USE_XATTRS

==> Installing mpifileutils
==> Fetching https://github.com/hpc/mpifileutils/releases/download/v0.7/mpifileutils-0.7.tar.gz
######################################################################## 100.0%
==> Staging archive: /root/HPE-Lustre/spack/var/spack/stage/mpifileutils-0.7-4fehgouyhkpet3tfogxriuwmdwx26jac/mpifileutils-0.7.tar.gz
==> Created stage in /root/HPE-Lustre/spack/var/spack/stage/mpifileutils-0.7-4fehgouyhkpet3tfogxriuwmdwx26jac
==> No patches needed for mpifileutils
==> Building mpifileutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j32' 'CFLAGS=-DDCOPY_USE_XATTRS'

31 errors found in build log:
     204      CC       libmfu_la-mfu_path.lo
     205      CC       libmfu_la-strmap.lo
     206      CC       libmfu_la-mfu_util.lo
     207      CC       libmfu_la-mfu_flist_chunk.lo
     208    In file included from mfu.h:48,
     209                     from mfu_util.c:21:
  >> 210    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     211    In file included from mfu_util.c:23:
     212    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype
     213    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
     214    In file included from mfu.h:48,
     215                     from mfu_path.c:23:
  >> 216    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     217    In file included from mfu.h:48In file included from mfu.h:48,
     218                     from mfu_io.c:21,
     219                     from mfu_param_path.c:21:
     220    :
  >> 221    mfu_flist.h:68:24:mfu_flist.h:68:24:  error: error: attr/xattr.h: No such file or direc
            toryattr/xattr.h: No such file or directory
     222
     223    mfu_util.c: In function 'mfu_memalign':
     224    mfu_util.c:110: warning: implicit declaration of function 'posix_memalign'
     225    mfu_util.c: In function 'mfu_strdupf':
     226    mfu_util.c:158: warning: conversion to 'size_t' from 'int' may change the sign of the r
            esult
     227    mfu_util.c:161: warning: conversion to 'size_t' from 'int' may change the sign of the r
            esult

     ...

     247    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
     248    In file included from mfu_flist_remove.c:41:
     249    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype
     250    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
     251    In file included from mfu.h:48,
     252                     from strmap.c:12:
  >> 253    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
  >> 254    make[3]: *** [libmfu_la-mfu_util.lo] Error 1
     255    make[3]: *** Waiting for unfinished jobs....
     256    In file included from mfu_flist_usrgrp.c:58:
     257    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype
     258    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
  >> 259    make[3]: *** [libmfu_la-mfu_io.lo] Error 1
     260    In file included from mfu_flist_io.c:57:
     261    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype
     262    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
     263    strmap.c: In function 'strmap_node_search':
     264    strmap.c:293: warning: cast discards qualifiers from pointer target type
     265    strmap.c:300: warning: cast discards qualifiers from pointer target type
     266    strmap.c:307: warning: cast discards qualifiers from pointer target type
     267    strmap.c:320: warning: cast discards qualifiers from pointer target type
     268    strmap.c:327: warning: cast discards qualifiers from pointer target type
     269    In file included from mfu.h:48,
     270                     from mfu_flist_sort.c:41:
  >> 271    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     272    In file included from mfu.h:48,
     273                     from mfu_flist_remove.c:42:
  >> 274    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
  >> 275    make[3]: *** [libmfu_la-mfu_param_path.lo] Error 1
     276    In file included from mfu.h:48,
     277                     from mfu_flist_usrgrp.c:59:
  >> 278    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     279    strmap.c: In function 'strmap_unset':
     280    strmap.c:770: warning: cast discards qualifiers from pointer target type
     281    In file included from mfu.h:48,
     282                     from mfu_flist_io.c:58:
  >> 283    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
  >> 284    make[3]: *** [libmfu_la-mfu_path.lo] Error 1
     285    In file included from mfu.h:48,
     286                     from mfu_flist_chunk.c:29:
  >> 287    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     288    mfu_flist_remove.c: In function 'remove_spread':
     289    mfu_flist_remove.c:314: warning: value computed is not used
     290    mfu_flist_io.c: In function 'list_elem_pack_size':
     291    mfu_flist_io.c:175: warning: unused parameter 'elem'
     292    In file included from mfu_flist_walk.c:57:
     293    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype

     ...

     310    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:69: warning: function declaration isn't a prototype
     311    /root/HPE-Lustre/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/dtcmp-1.1.0-m3ahriwkhk4ml
            uxk22wih7li6atie3su/include/dtcmp.h:72: warning: function declaration isn't a prototype
     312    mfu_flist_io.c: In function 'mfu_flist_read_cache':
     313    mfu_flist_io.c:758: warning: cast discards qualifiers from pointer target type
     314    In file included from mfu.h:48,
     315                     from mfu_flist_walk.c:58:
  >> 316    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     317    mfu_flist_io.c: In function 'write_cache_readdir_variable':
     318    mfu_flist_io.c:874: warning: cast discards qualifiers from pointer target type
     319    In file included from mfu.h:48,
     320                     from mfu_flist_copy.c:67:
  >> 321    mfu_flist.h:68:24: error: attr/xattr.h: No such file or directory
     322    mfu_flist_io.c: In function 'write_cache_readdir':
     323    mfu_flist_io.c:1014: warning: cast discards qualifiers from pointer target type
  >> 324    make[3]: *** [libmfu_la-strmap.lo] Error 1
     325    mfu_flist_io.c: In function 'write_cache_stat':
     326    mfu_flist_io.c:1159: warning: cast discards qualifiers from pointer target type
     327    mfu_flist_walk.c:166:5: warning: "HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC" is not defined
     328    mfu_flist_walk.c:174:7: warning: "HAVE_STRUCT_STAT_ST_MTIME_N" is not defined
     329    mfu_flist_walk.c:178:7: warning: "HAVE_STRUCT_STAT_ST_UMTIME" is not defined
     330    mfu_flist_walk.c:182:7: warning: "HAVE_STRUCT_STAT_ST_MTIME_USEC" is not defined

     ...

     337    mfu_flist_copy.c:134: warning: passing argument 1 of 'mfu_path_delete' from incompatibl
            e pointer type
     338    mfu_flist_walk.c:215: warning: unused parameter 'size1'In file included from mfu.h:48
     339    mfu_path.h:90: note: expected 'struct mfu_path **' but argument is of type 'const struc
            t mfu_path **'
     340    ,
     341                     from mfu_flist.c:64mfu_flist_walk.c:215: warning: unused parameter 'si
            ze2':
     342    mfu_flist.h:68:24:
  >> 343    error: attr/xattr.h: No such file or directory
     344    mfu_flist_copy.c:140: warning: passing argument 1 of 'mfu_path_delete' from incompatibl
            e pointer type
     345    mfu_path.h:90: note: expected 'struct mfu_path **' but argument is of type 'const struc
            t mfu_path **'
     346    mfu_flist_walk.c: In function 'reduce_fini':
     347    mfu_flist_walk.c:223: warning: unused parameter 'size'
     348    mfu_flist_walk.c: In function 'walk_getdents_process_dir':
     349    mfu_flist_walk.c:482: warning: conversion to 'int' from 'long int' may alter its value
     350    mfu_flist_copy.c: In function 'mfu_copy_xattrs':
     351    mfu_flist_copy.c:229: warning: implicit declaration of function 'llistxattr'
  >> 352    make[3]: *** [libmfu_la-mfu_flist_usrgrp.lo] Error 1
     353    mfu_flist_copy.c:279: warning: implicit declaration of function 'lgetxattr'
  >> 354    mfu_flist_copy.c:288: error: 'ENOATTR' undeclared (first use in this function)
  >> 355    mfu_flist_copy.c:288: error: (Each undeclared identifier is reported only once
  >> 356    mfu_flist_copy.c:288: error: for each function it appears in.)
     357    mfu_flist.c: In function 'list_elem_pack2_size':
     358    mfu_flist.c:114: warning: unused parameter 'elem'
     359    mfu_flist_copy.c:317: warning: implicit declaration of function 'lsetxattr'
  >> 360    make[3]: *** [libmfu_la-mfu_flist_remove.lo] Error 1
     361    mfu_flist_walk.c: In function 'walk_readdir_process_dir':
     362    mfu_flist_walk.c:667: warning: conversion to 'mode_t' from 'int' may change the sign of
             the result
     363    mfu_flist_walk.c: In function 'walk_readdir_create':
     364    mfu_flist_walk.c:713: warning: initialization discards qualifiers from pointer target t
            ype
  >> 365    make[3]: *** [libmfu_la-mfu_flist_sort.lo] Error 1
     366    mfu_flist_walk.c: In function 'walk_stat_create':
     367    mfu_flist_walk.c:805: warning: passing argument 1 of 'handle->enqueue' discards qualifi
            ers from pointer target type
     368    mfu_flist_walk.c:805: note: expected 'char *' but argument is of type 'const char *'
     369    mfu_flist_copy.c: In function 'mfu_copy_set_metadata':
     370    mfu_flist_copy.c:481: warning: unused parameter 'minlevel'
  >> 371    make[3]: *** [libmfu_la-mfu_flist_chunk.lo] Error 1
     372    mfu_flist_walk.c: In function 'mfu_flist_stat':
     373    mfu_flist_walk.c:1032: warning: too few arguments for format
  >> 374    make[3]: *** [libmfu_la-mfu_flist_io.lo] Error 1
     375    mfu_flist_copy.c: In function 'mfu_create_files':
     376    mfu_flist_copy.c:857: warning: passing argument 4 of 'mfu_create_file' discards qualifi
            ers from pointer target type
     377    mfu_flist_copy.c:736: note: expected 'struct mfu_param_path *' but argument is of type
            'const struct mfu_param_path *'
     378    mfu_flist.c: In function 'mfu_flist_remap':
     379    mfu_flist.c:1083: warning: passing argument 4 of 'map' discards qualifiers from pointer
             target type
     380    mfu_flist.c:1083: note: expected 'void *' but argument is of type 'const void *'
     381    mfu_flist_copy.c: In function 'mfu_copy_file_normal':
     382    mfu_flist_copy.c:947: warning: conversion to 'off_t' from 'uint64_t' may change the sig
            n of the result
     383    mfu_flist_copy.c:954: warning: conversion to 'off_t' from 'uint64_t' may change the sig
            n of the result
     384    mfu_flist.c: In function 'map_spread':
     385    mfu_flist.c:1165: warning: unused parameter 'args'
     386    mfu_flist_copy.c:1081: warning: conversion to 'off_t' from 'uint64_t' may change the si
            gn of the result
  >> 387    make[3]: *** [libmfu_la-mfu_flist_walk.lo] Error 1
  >> 388    make[3]: *** [libmfu_la-mfu_flist.lo] Error 1
  >> 389    make[3]: *** [libmfu_la-mfu_flist_copy.lo] Error 1
     390    make[3]: Leaving directory `/tmp/root/spack-stage/spack-stage-nCE_hq/mpifileutils-0.7/s
            rc/common'
  >> 391    make[2]: *** [all-recursive] Error 1
     392    make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-nCE_hq/mpifileutils-0.7/s
            rc'
  >> 393    make[1]: *** [all-recursive] Error 1
     394    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-nCE_hq/mpifileutils-0.7'
  >> 395    make: *** [all] Error 2

**Comment by Zubrania** [comment]
Finally  have  manged  with  the building  the tool

The strange  thing is  that  I am  getiing  strange at least for me results 

Please let me know If I can  send  private message to you about my  issue with dcp and my  environment

**Comment by dsikich** [comment]
@Zubrania yes, you can email me at sikich1@llnl.gov

**Comment by adammoody** [comment]
@Zubrania , based on this thread, it looks like you were able to get the tools building, so I'll close this particular issue.  Please feel free to open a new issue if you're hitting other problems.  Thanks.
